cs/src/core/Serializer.cs

```diff
@@ -76,10 +75,18 @@ public class Serializer<W>
         public Serializer(Type type, bool inlineNested)
         {
             var parser = new ObjectParser(type);
-            serialize = SerializerGeneratorFactory<object, W>.Create(
-                    (o, w, i) => serialize[i](o, w), type, inlineNested)
-                .Generate(parser)
-                .Select(lambda => lambda.Compile()).ToArray();
+
+            var firstPassAttribute = typeof(W).GetAttribute<FirstPassWriterAttribute>();
+            if (firstPassAttribute != null)
+            {
+                Type firstPassHelperType = typeof(TwoPassSerializerHelper<>).MakeGenericType(typeof(W), firstPassAttribute.Type);
```

Comment:
We should check that `W` implements `ITwoPassProtocolWriter` here so that we can fail fast with a good error message. Throwing `ArgumentException` seems reasonable to me. If we don't do this, then we'll get an `InvalidCastException` from `TwoPassSerializerHelper.Serialize` down the line.